### ISSUE TITLE
Implementation of a pipeline for producing seasonal Arctic Sea ice forecasts using IceNet

### ISSUE THREAD

**Issue by annefou** [bgruening/project-ideas#35]
Implementation of a pipeline for producing seasonal Arctic Sea ice forecasts using IceNet
---

**Supervisor:** @annefou @j34ni
**For degree:** Master 
**Status:** Open
**Keywords:** climate - deep-learning - probability - ERA5 

# Global Biological/Research context

Anthropogenic warming has led to an unprecedented year-round reduction in Arctic sea ice extent. This has far-reaching consequences for indigenous and local communities, polar ecosystems, and global climate, motivating the need for accurate seasonal sea ice forecasts. IceNet is a probabilistic, deep learning sea ice forecasting system. The system has been trained on climate simulations and observational data to forecast the next 6 months of monthly-averaged sea ice concentration maps. IceNet advances the range of accurate sea ice forecasts, outperforming a state-of-the-art dynamical model in seasonal forecasts of summer sea ice, particularly for extreme sea ice events. This step-change in sea ice forecasting ability brings us closer to conservation tools that mitigate risks associated with rapid sea ice loss.

# Project context

The [Environmental Data Science Book) ](https://the-environmental-ds-book.netlify.app/welcome.html) initiative is a community driven initiative. The Environmental Data Science book is a living, open and community-driven online resource to showcase and support the publication of data, research and open-source tools for collaborative, reproducible and transparent Environmental Data Science. Developing Galaxy Tools from the developed Jupyter Notebook will male our initiative more accessible and inclusive, allowing to produce fully reproducible workflows either through command line or via the Galaxy Graphical Interface.

# Objectives of the project

The main objective of this project is to implement Galaxy Tools for seasonal sea-ice forecasting system with IceNet using an existing Jupyter Notebook, and to develop a training for illustrating to the Galaxy community the required steps for running it.

# Proposed agenda for the project

1. Get familiar with  the existing [Jupyter Notebook](https://github.com/Environmental-DS-Book/polar-modelling-icenet/blob/main/polar-modelling-icenet.ipynb) (this notebook demonstrates the usage of IceNet). 
2. Get familiar with Galaxy
3. Identify the different steps of the workflow and develop the corresponding Galaxy Tools
4.  Develop a training for the GTN

# Prerequisites

- [Required] Basic knowledge of deep-learning
- [Required] Basic knowledge of Linux and version control (Git)
- [Required] Python
- Data analysis through Jupyter Notebook

# Further reading and useful links

- Further details can be found in the Nature Communications paper [Seasonal Arctic sea ice forecasting with probabilistic deep learning](https://www.nature.com/articles/s41467-021-25257-4).
- The rendered Jupyter notebook that will be the basis for this work can be found [here](https://the-environmental-ds-book.netlify.app/gallery/modelling/polar-modelling-icenet/polar-modelling-icenet.html).

**Comment by tom-andersson** [comment]
Hey @annefou, my colleague @acocac sent me this -- this is cool! Just FYI, there is now a `pip install icenet` option available, see: https://github.com/icenet-ai/icenet. Documentation is a WIP, but cc lead developer @JimCircadian in case you are interested.

**Comment by acocac** [comment]
@tom-andersson @JimCircadian (cc'Ed @aldenc @scotthosking) just a head-up this project is going really well. In short, @vanessa-tamara is porting the IceNet notebook published in the EDS book to the [Galaxy platform](https://usegalaxy.eu/). Galaxy is an open, web-based platform for accessible, reproducible, and transparent computational biological research, but has been extended to other domains, including Climate and Earth System sciences (see for instance [here](https://training.galaxyproject.org/training-material/topics/climate/)). I suggest then including Vanessa's master work in the IceNet impact stories.

**Comment by JimCircadian** [comment]
Thanks @acocac and @vanessa-tamara! We are exhibiting at AI UK 2023 on 21-22 March so would be awesome to refer / link to this as an impact story when we are there. Point us at some resources if they're publicly available and we'll be sure to mention this!

**Comment by annefou** [comment]
@JimCircadian you can reference this presentation: https://doi.org/10.5281/zenodo.7723371 (corresponding google slides at https://docs.google.com/presentation/d/1mCfH7kmyUQ7aCoKk3jeFtSq6SM2O0XQVSimvpoaYAsE/edit?usp=sharing). Hope it helps.

**Comment by vanessa-tamara** [comment]
Hi @tom-andersson  @JimCircadian , i'm currently preparing the galaxy tool for forecasting and i have some questions about the linear trend data. Is it correct, that as an input for the network one also needs the sea ice concentration linear trend data for 6 month in the future even if i take the pretrained icenet models from https://the-environmental-ds-book.netlify.app/notebooks/gallery/ac327c3a-5264-40a2-8c6e-1e8d7c4b37ef/notebook.html?highlight=icenet#download-pretrained-icenet-models ? 
And for preprocessing the climate variable data (anomalies) would it be sufficient to take the json file with the precomputed normalization parameters from https://zenodo.org/record/5516869/#.ZBt3HcKZM2w in order to avoid having to download 35 years of linear trend data to compute the anomalies? 
Kind regards, 
Vanessa

**Comment by tom-andersson** [comment]
Hi @vanessa-tamara, yes you will need the linear trend forecasts to run IceNet because this is one of its input variables -- see [Supplementary Table 2](https://www.nature.com/articles/s41467-021-25257-4#Sec22:~:text=and%20institutional%20affiliations.-,Supplementary%20information,-Supplementary%20Information%20for) of the seasonal forecasting paper. I'm afraid you will have to download all the SIC data to compute these linear trend forecasts. There is code provided for this in the [icenet-paper](https://github.com/tom-andersson/icenet-paper) repository.

I don't think the JSON file contains the precomputed normalisation parameters, right? In the Zenodo released with @acocac's EDS notebook, `dataset1.zip` contains some normalised data that he produced, which you can use to run the model over 2020. Are you not using this? Unfortunately there is no way to avoid downloading the 35 years of data if you wish to normalise the data yourself -- the data pipeline for the seasonal forecasting model is optimised for research rather than operations, whereas the [`icenet` Python package](https://github.com/icenet-ai/icenet) for daily forecasting (which @JimCircadian has developed) is suitable for operational purposes and has much more streamlined data pipeline functionality.

**Comment by vanessa-tamara** [comment]
Hi, @tom-andersson alright thanks a lot!
I thought so because of: 
"Normalisation parameters computed over the observational training data are stored in a JSON file at data/network_datasets/<dataset_name>/norm_params.json so that they are only computed once." in preproc_icenet_data.py.. Is this another file?
Yes, we wanted to run the model not only for 2020..

**Comment by tom-andersson** [comment]
Hi @vanessa-tamara, ah right, yes, that is a separate JSON config file. The JSON file that @acocac put on the Zenodo is the data loader configuration file, specifying the input variables etc.

Even if you had our pre-computed `norm_params.json` file, you would still need to download all the ERA5 data back to 1979 to compute the monthly climatologies in order to compute the anomaly variables. Sorry that this is inconvenient.

@JimCircadian might have suggestions if you wish to switch to use the `icenet` Python package rather than the old `icenet-paper` research codebase.

**Comment by vanessa-tamara** [comment]
Hi @tom-andersson , norm_params.json was also uploaded in the zenodo repository (in dataset1.zip), I thought the mean and std are maybe only computed once for the training months (so it would always be the same? ), like the function says:

normalise_array_using_all_training_months(self, da, minmax=False,
                                                  mean=None, std=None,
                                                  min=None, max=None):

        """
        Using the *training* months only, compute the mean and
        standard deviation of the input raw satellite DataArray (`da`)
        and return a normalised version. If minmax=True,
        instead normalise to lie between min and max of the elements of `array`.

For the environmental data book example the obs_train_dates were from 1980-1-1 to 2011-6-1 according to the config file and it forecasts 2020. So it's not enough to consider those?

**Comment by JimCircadian** [comment]
@vanessa-tamara the [icenet](https://github.com/icenet-ai/icenet) operational library deals with this slightly differently, in that if you're using pre-trained models you can reference the climatologies and normalisation parameters from the training set for predictions of that model, as well as use the linear trend file associated to the network. 

For the latter, the linear trend file is computed during pre-processing of the training data when that's prepared, which means it's effectively shareable alongside the networks. This allows users who are only predicting to symlink/share the training dataset without having to regenerate it. The linear trend files themselves are large but this does make things a little easier when people aren't interested in retraining the networks. However, there's no real way of getting around either needing a pre-prepared linear trend file (as icenet-paper requires) or having to download and prepare one yourself, regardless of which library you're using. If you're committed to basing your workflow on the icenet-paper repository, I think you'll need to follow @tom-andersson's instructions above for generating it. 

In terms of the climatologies, we do store these so that they're reusable in the new pipeline, but I don't have ones that will correspond to the icenet-paper published models for monthly/seasonal forecasts. [We did discuss and open a ticket for use of monthly data](https://github.com/icenet-ai/icenet/issues/22), which will open the door also to generating these with the operational pipeline, but that's not yet been progressed.

**Comment by tom-andersson** [comment]
@vanessa-tamara Ah yes, I didn't realise you were referring to `norm_params.json`, thanks for clarifying. Yes, these are only computed once for a given variable. If the file already has the normalisation parameters, these will not be computed when normalising the data.

You will however still need to download the full training period in order to compute the monthly climatology NetCDFs for the 8 anomaly variables. You will also need the full sea ice concentration dataset from OSI-SAF in order to compute the linear trend forecasts which are input to IceNet (there is no way around this, as @JimCircadian says).

Note that a ~40 year NetCDF file of a monthly-averaged variable on the NH EASE2 grid is around 730 MB and a monthly climatology file is only ~18 MB. One thing you could do is deposit the climatologies on Zenodo so that they don't need to be computed as part of the notebook (you will need to check ERA5 licensing terms though). The reason @acocac put his pre-normalised network dataset on Zenodo is that it avoids this annoying data-wrangling step which is currently causing you issues. The `icenet-paper` codebase wasn't really designed around the end-to-end pipeline fitting into a single Jupyter notebook with a small memory footprint, so I think you will need to bypass some steps if you don't want the notebook to take days to run and require 10s of GBs of storage.

Sorry I won't be able to assist further until late April because I'm about to go on annual leave. However, I hope this has clarified things and I look forward to seeing how this turns out!

**Comment by bgruening** [comment]
Very good work everyone! Thanks a lot!
